Fixes #11783 - MS providers: add compat for ruby2.0

helge000 opened the pull request:
If folks have also puppet installed on there machine running the smart proxy, there is no need to install ruby 1.9.3.
Instead, puppet3 and puppet4 ruby versions be used with this simple modification for the `win32/open3` gem requirement in ruby 1.9.3
DANIEL HELGENBERGER daniel.helgenberger@m-box.de

helge000 committed:
Fixes #11783 - MS providers: add compat for ruby2.0
--
DANIEL HELGENBERGER <daniel.helgenberger@m-box.de>

dmitri-d commented:
[test]

helge000 commented:
Changed commit msg

domcleal commented:
I've no idea which is more correct, but we also have this open PR which contains the following similar line: https://github.com/theforeman/smart-proxy/pull/294/files#diff-08c4251c1dbcdc5c20755615254654eaR2

dmitri-d commented:
I'm finally getting around to creating a Windows VM, will test this and older windows-specific PRs. Will keep y'all updated.

helge000 commented:
@domcleal, I think you mean line 2 in `modules/dhcp/providers/server/native_ms.rb` of this PR:
`require 'win32/open3' if RUBY_PLATFORM =~ /mingw/`
does actually nothing as the native providers are only loaded in windows anyway?
That said, I do basically the same thing ;).

Btw, another issue not handled by my PR is ruby 1.9 in `bundler.d/windows.rb`; `win32/open3` will not be installed by bundler. Excuse my ignorance, Is Ruby 1.8 still supported for smart proxy? If so, I should change my PR to reflect that;
`RbConfig::CONFIG['target_os'] == 'mingw32' && !(RUBY_VERSION =~ /1.9/) && !(RUBY_VERSION =~ /2.0/) ? 'win32/open3' : 'open3'`

to something like:
`RbConfig::CONFIG['target_os'] == 'mingw32' && !(RUBY_VERSION =~ /1.(8|9)/) && !(RUBY_VERSION =~ /2.0/) ? 'win32/open3' : 'open3'`

dmitri-d commented:
@helge000: yes, we support ruby 1.8.7.

domcleal commented:
> @domcleal, I think you mean line 2 in modules/dhcp/providers/server/native_ms.rb of this PR:
> require 'win32/open3' if RUBY_PLATFORM =~ /mingw/

Yes, I did, sorry.

> does actually nothing as the native providers are only loaded in windows anyway?

I've just merged a partial refactoring of these classes that lets us load them on non-Windows platforms, just for the purposes of testing.

Please ignore my original comment actually, I see now that you're changing the behaviour based on Ruby version rather than platform as I thought.

helge000 committed:
Fixes #11783 - Re-added missing Ruby 1.8 support
--
DANIEL HELGENBERGER <daniel.helgenberger@m-box.de>

helge000 commented:
@witlessbird, thanks for the info. I changed the relevant regex to hopefully re-enable ruby 1.8.

dmitri-d commented:
@helge000: there were some recent changes in develop branch, you'll need to rebase your PR. Also, a few of us had a conversation regarding support for ruby 1.8 on Windows and came to a conclusion it isn't worth the effort.

Since win32-open3 gem was 1.8-specific, you can replace conditional requires you introduced to `require 'open3'`. Apologies for going back and forth on 1.8 support.

dmitri-d commented on bundler.d/windows.rb:
```diff
@@ -2,4 +2,5 @@
   gem 'highline', :platform => :mingw
   gem 'win32-service', :platforms => :mingw
   gem "win32-open3", :platforms => :mingw_18
+  gem "open3", :platforms => :mingw_20
```
You don't need this line, `open3` is available in the baseline ruby starting with 1.9.3 (and there's no gem called 'open3'. Chances are you are on x64_mingw_20, that's why bundler didn't complain).

dmitri-d commented on bundler.d/windows.rb:
```diff
@@ -2,4 +2,5 @@
   gem 'highline', :platform => :mingw
   gem 'win32-service', :platforms => :mingw
   gem "win32-open3", :platforms => :mingw_18
```
You can kill this line.

helge000 commented:
@witlessbird np, i'll do the rebase, but one question: If we you drop ruby 1.8 in windows, I'd change all this to `require open3` directly and remove refs to `win32-open3`?

This goes for bundler as well...

dmitri-d commented:
@helge000: nods, you can replace requires to `require "open3"`. You can also remove `win32-open3` dependency from the windows gem group. You need to remove `open3` gem dependency too, as there's no such gem (pls. see my comment above)

helge000 committed:
Rebase upsteam/develop; fix: bundle install on windows
--
DANIEL HELGENBERGER <daniel.helgenberger@m-box.de>

helge000 committed:
fiixes #11783 - rebase to upstream. Change remove win32/open3; fix
bundler for windows
Merge remote-tracking branch 'upsteam/develop' into ms-ruby2.0

helge000 commented:
@witlessbird: Ok, did the rebase; sorry about the typo. I still leave `wopen3` [here](https://github.com/helge000/smart-proxy/blob/develop/bundler.d/windows.rb#L4). 

Instead of the regex I just use `if RUBY_PLATFORM =~ /mingw/` now; as already merged also in [dns](https://github.com/helge000/smart-proxy/blob/ms-ruby2.0/modules/dns_dnscmd/dns_dnscmd_main.rb#L2).

Further, I removed krb5 form bundler and puppet/augeas with `if RUBY_PLATFORM =~ /ruby/` in windows. I just tested with a clean VM, ruby 2.0 &  DevKit. `bundle install` runs now without errors and `ruby bin/smart-proxy` starts up out of the box (provided some config)

dmitri-d commented on bundler.d/windows.rb:
```diff
@@ -1,5 +1,5 @@
   gem 'highline', :platform => :mingw
   gem 'win32-service', :platforms => :mingw
-  gem "win32-open3", :platforms => :mingw_18
+  gem 'wopen3', '>= 0.3', :platforms => :mingw
```
I haven't tested smart-proxy with wopen3 gem, and I don't think there's a need for it: modern rubies' implementation of Open3 library is good enough for us.

helge000 replied:
Right, this gem is included and bundler will know about it and not install it. I was unsure and kept it as it does not any harm. You think I should remove it?

dmitri-d commented on bundler.d/krb5.rb:
```diff
@@ -1,4 +1,4 @@
-  gem 'rkerberos', '>= 0.1.1'
-  gem 'gssapi'
+  gem 'rkerberos', '>= 0.1.1', :platform => :ruby
+  gem 'gssapi', :platforms => :ruby
```
I don't like limiting the platform to `ruby` here. While bundler won't complain during `bundler install` on Windows, the resulting installation isn't going to be functional. I prefer that users using bundler-based installs explicitly skipped groups, ie: `bundler install --without krb5`. I'll make sure the install docs get updated if this information isn't there.

dmitri-d commented on bundler.d/puppet.rb:
```diff
@@ -1,4 +1,4 @@
 group :puppet, :puppetca do
   gem 'puppet', '< 4.0.0'
-  gem 'ruby-augeas', :require => 'augeas'
+  gem 'ruby-augeas', :require => 'augeas', :platform => :ruby
```
ditto here.

helge000 replied:
Ok, I see your point. Missed the comment above. I move it to doc

dmitri-d commented on modules/dhcp/providers/server/native_ms.rb:
```diff
@@ -1,5 +1,5 @@
 require 'checks'
-require 'win32/open3' if RUBY_PLATFORM =~ /mingw/
+require 'open3' if RUBY_PLATFORM =~ /mingw/
```
No need for a conditional `require` here, `Open3` is available on all platforms (except mingw_18, which we don't support on Windows).

helge000 commented:
@domcleal I 'fixed' the `require` statements in windows for puppet by adding `if RUBY_PLATFORM =~ /ruby/`. 
I just remembered: the issue should also apply to Kerberos  and does not. So I think it might be introduced with the puppet providers?

dmitri-d commented on modules/dns_dnscmd/dns_dnscmd_main.rb:
```diff
@@ -1,5 +1,5 @@
 require 'resolv'
-require 'win32/open3'
+require 'open3' if RUBY_PLATFORM =~ /mingw/
```
ditto.

helge000 replied:
I added `if` line only because it was already merged...

dmitri-d replied:
Not sure what you mean by that.

dmitri-d commented on modules/puppet_proxy/initializer.rb:
```diff
@@ -1,4 +1,4 @@
-require 'puppet'
+require 'puppet' if RUBY_PLATFORM =~ /ruby/
```
Again, I don't think this is useful: the module isn't going to work without this dependency, this will only hide the error which otherwise would disable the module on startup.

helge000 replied:
Just tested this again. Removing the conditionals form puppt-proxy leaves me with:

```
c:/smart-proxy/modules/puppet_proxy/initializer.rb:1:in `require': cannot load such file -- puppet (LoadError)
        from c:/smart-proxy/modules/puppet_proxy/initializer.rb:1:in `<top (required)>'
        from c:/smart-proxy/modules/puppet_proxy/puppet_plugin.rb:1:in `require'
        from c:/smart-proxy/modules/puppet_proxy/puppet_plugin.rb:1:in `<top (required)>'
        from c:/smart-proxy/modules/puppet_proxy/puppet.rb:1:in `require'
        from c:/smart-proxy/modules/puppet_proxy/puppet.rb:1:in `<top (required)>'
        from c:/smart-proxy/lib/smart_proxy.rb:63:in `require'
        from c:/smart-proxy/lib/smart_proxy.rb:63:in `<module:Proxy>'
        from c:/smart-proxy/lib/smart_proxy.rb:38:in `<top (required)>'
        from C:/Ruby200-x64/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in `require'
        from C:/Ruby200-x64/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in `require'
        from bin/smart-proxy:5:in `<main>'
```

dmitri-d commented on modules/puppet_proxy/puppet_config.rb:
```diff
@@ -1,4 +1,4 @@
-require 'augeas'
+require 'augeas' if RUBY_PLATFORM =~ /ruby/
```
ditto.

helge000 replied:
I do not like it either; though removing it currently prevents smart proxy from running in windows

dmitri-d replied:
Hmm, I'm quite certain this isn't the case. Make sure puppet, puppetca, and bmc modules are disabled in their respective config files.

helge000 replied:
right you are; using the correct bundle install command does fix this. I revert all this back

helge000 replied:
If I remove the if:

```
c:/smart-proxy/modules/puppet_proxy/puppet_config.rb:1:in `require': cannot load such file -- augeas (LoadError)
        from c:/smart-proxy/modules/puppet_proxy/puppet_config.rb:1:in `<top (required)>'
        from c:/smart-proxy/modules/puppet_proxy/runtime_configuration.rb:2:in `require'
        from c:/smart-proxy/modules/puppet_proxy/runtime_configuration.rb:2:in `<top (required)>'
        from c:/smart-proxy/modules/puppet_proxy/ssl_configuration_validator.rb:1:in `require'
        from c:/smart-proxy/modules/puppet_proxy/ssl_configuration_validator.rb:1:in `<top (required)>'
        from c:/smart-proxy/modules/puppet_proxy/puppet_plugin.rb:2:in `require'
        from c:/smart-proxy/modules/puppet_proxy/puppet_plugin.rb:2:in `<top (required)>'
        from c:/smart-proxy/modules/puppet_proxy/puppet.rb:1:in `require'
        from c:/smart-proxy/modules/puppet_proxy/puppet.rb:1:in `<top (required)>'
        from c:/smart-proxy/lib/smart_proxy.rb:63:in `require'
        from c:/smart-proxy/lib/smart_proxy.rb:63:in `<module:Proxy>'
        from c:/smart-proxy/lib/smart_proxy.rb:38:in `<top (required)>'
        from C:/Ruby200-x64/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in `require'
        from C:/Ruby200-x64/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in `require'
        from bin/smart-proxy:5:in `<main>'
```

dmitri-d replied:
This is a bug I introduced in https://github.com/theforeman/smart-proxy/pull/317. Looking into this, but assume that if a module is disabled its dependencies will not be loaded.

dmitri-d commented:
@domcleal, @GregSutcliffe: thoughts on limiting puppet and augeas dependencies to `ruby` platform only? I don't like the idea, pls. see my comment.

helge000 commented on bundler.d/windows.rb:
```diff
@@ -1,5 +1,5 @@
 group :windows do
   gem 'highline', :platform => :mingw
```
Then we should remove `:platform => :mingw` here as well?

dmitri-d replied:
Yes, let's do that.

helge000 committed:
assume if module is disabled its dependencies will not be loaded

DANIEL HELGENBERGER <daniel.helgenberger@m-box.de>

dmitri-d commented:
@helge000: https://github.com/theforeman/smart-proxy/pull/321 fixes the issue with puppet dependencies loading that you discovered (thanks!).

helge000 committed:
fixes #11783  - revert bundler changes and conditionals
Merge remote-tracking branch 'upsteam/develop' into ms-ruby2.0

helge000 commented:
@witlessbird: happy to help out unintentionally ;)
Rebased the branch and reverted bundler chnages

dmitri-d commented:
Could you squash your commits pls.?

dmitri-d commented:
[test]

dmitri-d commented:
Also, did you use https://github.com/theforeman/smart-proxy/pull/321 to test the changes? If so, could you comment in the PR re: whether changes worked.

helge000 commented:
@witlessbird just a checkup, anything left to do on my part?
For the record, #321 was tested; commented [here](https://github.com/theforeman/smart-proxy/pull/321#issuecomment-144438697)

dmitri-d commented:
@helge000: please squash the commits. There's also a comment here: https://github.com/theforeman/smart-proxy/pull/318/files#diff-aede4453fda32f153191fffeb8d474dcR2 that I don't understand.

dmitri-d commented:
[test]

helge000 committed:
Fixes #11783 - Merge remote-tracking branch 'upsteam/develop' into ms-ruby2.0

helge000 commented:
> @helge000: please squash the commits. There's also a comment here: https://github.com/theforeman/smart-proxy/pull/318/files#diff-aede4453fda32f153191fffeb8d474dcR2 that I don't understand.

@witlessbird - sorry, my bad; forgot about comment. What I meant though was it was there when I merged upstream/develop so I figured wrongly this should stay (I think was merged with #294 ). I removed these conditionals now.

dmitri-d commented:
[test]

dmitri-d commented:
Merged in at 5f36f2e. Thanks, @helge000!
